pkgs/top-level/by-name-overlay.nix

```diff
@@ -9,36 +9,23 @@ let
   # Because of Nix's import-value cache, importing lib is free
   lib = import ../../lib;
 
-  inherit (builtins)
-    readDir
-    ;
-
   inherit (lib.attrsets)
-    mapAttrs
-    mapAttrsToList
     mergeAttrsList
+    mapAttrsRecursive
+    mapAttrsToList
     ;
 
-  # Package files for a single shard
-  # Type: String -> String -> AttrsOf Path
-  namesForShard = shard: type:
-    if type != "directory" then
-      # Ignore all non-directories. Technically only README.md is allowed as a file in the base directory, so we could alternatively:
-      # - Assume that README.md is the only file and change the condition to `shard == "README.md"` for a minor performance improvement.
-      #   This would however cause very poor error messages if there's other files.
-      # - Ensure that README.md is the only file, throwing a better error message if that's not the case.
-      #   However this would make for a poor code architecture, because one type of error would have to be duplicated in the validity checks and here.
-      # Additionally in either of those alternatives, we would have to duplicate the hardcoding of "README.md"
-      { }
-    else
-      mapAttrs
-        (name: _: baseDirectory + "/${shard}/${name}/package.nix")
-        (readDir (baseDirectory + "/${shard}"));
+  inherit (lib.filesystem)
+    packageFilesFromDirectory
+    ;
 
-  # The attribute set mapping names to the package files defining them
-  # This is defined up here in order to allow reuse of the value (it's kind of expensive to compute)
-  # if the overlay has to be applied multiple times
-  packageFiles = mergeAttrsList (mapAttrsToList namesForShard (readDir baseDirectory));
+  # Read packages from `baseDirectory` and then strip the shard names (the
+  # outermost attributes).
+  packagePaths =
+    mergeAttrsList
+      (mapAttrsToList
+        (_shard: packages: packages)
+        (packageFilesFromDirectory { directory = baseDirectory; }));
```

Comment:
As already mentioned before, I don't think anything should be changed in this file. Yes I did ask to make it be shared with `pkgs/by-name` originally, but I don't think that should be done anymore. It's distinctly different that just having `packageFilesFromDirectory` for this very use case doesn't make any sense, especially when half of its functionality can't be used here anyways.

And considering that, we don't need this separate function.